### PR TITLE
Fix crashing on json serialization

### PR DESCRIPTION
**Pull request by TheAngryByrd**
### <samp>🤖 Generated by Copilot at 363f2ce</samp>

The pull request enhances the error handling and robustness of the LanguageServerProtocol library and its tests, using StreamJsonRpc types and methods and following the Fantomas style. It also updates the fantomas-tool package name and version in `dotnet-tools.json`.

### <samp>🤖 Generated by Copilot at 363f2ce</samp>

> _To avoid any naming mishaps_
> _We updated the package `fantomas`_
> _We also revised_
> _The `Server` and `Client` sides_
> _And added some tests and new RPCs_

:package::wrench::white_check_mark:

### <samp>🤖 Generated by Copilot at 363f2ce</samp>

*  Update fantomas-tool package name and version to fantomas 6.2.2 in `.config/dotnet-tools.json` ([link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-7afd3bcf0d0c06d6f87c451ef06b321beef770ece4299b3230bb280470ada2f6L11-R12))
*  Add open statements for StreamJsonRpc, JsonRpc, and StreamJsonRpc.Protocol namespaces in `src/LanguageServerProtocol.fs` ([link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-e8b1cdcb9bc744e4631def23ffbb4516012e5217882e1a32eb5e9d0cc68e56c9R16-R17))
*  Add (|Flatten|) active pattern and flatten function to handle AggregateExceptions in Server module ([link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-e8b1cdcb9bc744e4631def23ffbb4516012e5217882e1a32eb5e9d0cc68e56c9L85-R126))
*  Use flatten function in IsFatalException and CreateErrorDetails methods of defaultRpc function ([link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-e8b1cdcb9bc744e4631def23ffbb4516012e5217882e1a32eb5e9d0cc68e56c9L273-R307))
*  Reformat start function in Server module to break line after generic type parameters and remove extra parentheses ([link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-e8b1cdcb9bc744e4631def23ffbb4516012e5217882e1a32eb5e9d0cc68e56c9L273-R307))
*  Change catch clauses from using vertical bar to arrow in Client module ([link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-e8b1cdcb9bc744e4631def23ffbb4516012e5217882e1a32eb5e9d0cc68e56c9L328-R363), [link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-e8b1cdcb9bc744e4631def23ffbb4516012e5217882e1a32eb5e9d0cc68e56c9L456-R490), [link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-e8b1cdcb9bc744e4631def23ffbb4516012e5217882e1a32eb5e9d0cc68e56c9L475-R509), [link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-e8b1cdcb9bc744e4631def23ffbb4516012e5217882e1a32eb5e9d0cc68e56c9L560-R593))
*  Reformat MailboxProcessor.Start function call and loop function in mutable function in Client module ([link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-e8b1cdcb9bc744e4631def23ffbb4516012e5217882e1a32eb5e9d0cc68e56c9L427-R475))
*  Add open statements for Nerdbank.Streams, StreamJsonRpc, System, System.Reflection, and StreamJsonRpc.Protocol namespaces in `tests/StartWithSetup.fs` ([link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-68fe8350fe78602fb8fe7b610e6b7d0ad0314bb987f7f4cc2873a65c58278c70L8-R19))
*  Add Foo type and "lol" RPC method for testing JsonSerializationError handling in StartWithSetup module ([link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-68fe8350fe78602fb8fe7b610e6b7d0ad0314bb987f7f4cc2873a65c58278c70L22-R31), [link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-68fe8350fe78602fb8fe7b610e6b7d0ad0314bb987f7f4cc2873a65c58278c70L33-R46))
*  Add test case for JsonSerializationError handling in tests value in StartWithSetup module ([link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-68fe8350fe78602fb8fe7b610e6b7d0ad0314bb987f7f4cc2873a65c58278c70L33-R46))
*  Reformat test case for multiple start ups in tests value in StartWithSetup module ([link](https://github.com/ionide/LanguageServerProtocol/pull/55/files?diff=unified&w=0#diff-68fe8350fe78602fb8fe7b610e6b7d0ad0314bb987f7f4cc2873a65c58278c70L51-R104))

### CODE REVIEW

File: src/LanguageServerProtocol.fs [hunk 1]
```diff
@@ -82,14 +84,34 @@ module Server =
...
+
 
   /// The default RPC logic shipped with this library. All this does is mark LocalRpcExceptions as non-fatal
   let defaultRpc (handler: IJsonRpcMessageHandler) =
     { new JsonRpc(handler) with
         member this.IsFatalException(ex: Exception) =
           match ex with
-          | :? LocalRpcException -> false
-          | _ -> true }
+          | Flatten(:? LocalRpcException | :? JsonSerializationException) -> false
```
Comment (TheAngryByrd) [inline src/LanguageServerProtocol.fs:104]: Don't crash on `JsonSerializationException`

File: src/LanguageServerProtocol.fs [hunk 2]
```diff
@@ -82,14 +84,34 @@ module Server =
...
+          | _ -> true
+
+        member this.CreateErrorDetails(request: JsonRpcRequest, ex: Exception) =
+          let isSerializable = this.ExceptionStrategy = ExceptionProcessing.ISerializable
+
+          match ex with
+          | Flatten(:? JsonSerializationException as ex) ->
+            let data: obj = if isSerializable then ex else CommonErrorData(ex)
+            JsonRpcError.ErrorDetail(Code = JsonRpcErrorCode.ParseError, Message = ex.Message, Data = data)
```
Comment (TheAngryByrd) [inline src/LanguageServerProtocol.fs:113]: Create parse error for `JsonSerializationException` exception

File: tests/StartWithSetup.fs [hunk 3]
```diff
@@ -5,68 +5,101 @@ open System.IO.Pipes
...
+        inputWriter2.Write(requestWithContentLength (shutdownRequest))
+        inputWriter2.Write(requestWithContentLength (exitRequest))
 
         do! server1Async
         do! server2Async
       }
-    ]
\ No newline at end of file
+
+      testCaseAsync "Handle JsonSerializationError gracefully"
```
Comment (TheAngryByrd) [inline tests/StartWithSetup.fs:72]: Main test for this

**Comment by TheAngryByrd** [comment]
https://github.com/fsharp/FsAutoComplete/issues/1186

**Comment by TheAngryByrd** [comment]
I just realized we override [creation of `JsonRpc`](https://github.com/fsharp/FsAutoComplete/blob/be075d8809302ae62b454bce4bfcb721100bcde8/src/FsAutoComplete/LspServers/AdaptiveFSharpLspServer.fs#L3110) in FSAC already to do tracing.  I'll leave it up to you if you want this PR @baronfel since we'll still have to do a port over there anyway.

**Comment by baronfel** [comment]
Yeah, let's have it in both places so users of this  library have the correct behavior.

**Pull request merged**